### PR TITLE
chore(linting): move to eslint

### LINKED ISSUE #245: Facilitate EsLint
As TSLint became deprecated, ESLint should be used within the project

### PR DESCRIPTION
**Pull request by Retro64**
close #245

### CODE REVIEW

File: src/mock-tooling/index.ts [hunk 1]
```diff
@@ -24,7 +24,7 @@ const mockAccessTokenEndpoint = (options: MockOptions): nock.Scope => {
 
   const parsedUrl = parseUrlOrThrow(options);
 
-  return nock(`${parsedUrl.protocol}//${parsedUrl.host}`)
+  return nock(`${parsedUrl.protocol as string}//${parsedUrl.host as string}`)
```
[the remark this answers is not in the document]
Comment (Retro64) [reply src/mock-tooling/index.ts:27]: string | null

**Comment by Retro64** [comment]
:+1:

File: src/oauth-tooling.ts [hunk 2]
```diff
@@ -285,7 +286,10 @@ const getCredentials = (options: OAuthConfig): Promise<any>[] => {
     }
   }
 
-  return getUserData ? [getClientData, getUserData] : [getClientData];
+  return Promise.all([getUserData, getClientData])
```
Comment (ISO50) [inline src/oauth-tooling.ts:289]: This change does not seem to return the same data any more.

File: test/unit/mock-tooling.ts [hunk 3]
```diff
@@ -65,7 +66,8 @@ describe('mock-tooling', () => {
 
   describe('mockTokeninfoEndpointWithErrorResponse', () => {
     it('should throw error, if url is empty', () => {
-      return expect(() => mockTooling.mockTokeninfoEndpointWithErrorResponse({ url: '' }, HttpStatus.BAD_REQUEST)).to.throw(Error);
+      return expect(() =>
```
Comment (ISO50) [inline test/unit/mock-tooling.ts:69]: I think instead of `\n` it's better to get rid of the function body.

File: test/unit/express-tooling/authentication-middleware.ts [hunk 4]
```diff
@@ -39,7 +39,9 @@ describe('express tooling', () => {
       };
 
       // then
-      expect(() => { authenticationMiddleware(config); }).to.throw(TypeError);
+      expect(() => {
```
Comment (ISO50) [inline test/unit/express-tooling/authentication-middleware.ts:42]: I am not sure if this is more readable.
I suggest to leave the arrow function on one line in this case. Maybe just leave it as it was.

```
expect(
  () => { authenticationMiddleware(config); }
).to.throw(TypeError);
```

File: src/utils.ts [hunk 5]
```diff
@@ -61,8 +61,8 @@ const getHeaderValue = (req: Request, fieldName: string): string | undefined =>
   const headerValue = req && req.headers[fieldName];
 
   const normalizedHeaderValue = Array.isArray(headerValue) ?
-                                  headerValue.join(' ') :
-                                  headerValue;
+    headerValue.join(' ') :
+    headerValue;
```
Comment (ISO50) [inline src/utils.ts:65]:
```ts
  const normalizedHeaderValue = 
    Array.isArray(headerValue) ?
      headerValue.join(' ') :
      headerValue;
```

File: src/express-tooling.ts [hunk 6]
```diff
@@ -112,25 +118,25 @@ const requireScopesMiddleware: requireScopesMiddleware =
...
+        })
+        .then(isAllowed => {
+          if (isAllowed) { // precedenceFunction grants access
+            logOrNothing.debug('PrecedenceFunction grants access');
+            nextFunction();
+          } else if (validateScopes(request, scopes)) {  // scope validation grants access
+            logOrNothing.debug('Scopes validated successfully');
+            nextFunction();
+          } else {
+            logOrNothing.warn(`Scope validation failed for ${scopes.join(' ')}`);
```
Comment (ISO50) [inline src/express-tooling.ts:136]: I suggest `, ` as separator, it's closer to the standard array formatter i.e. `[1, 2, 4]`

File: src/express-tooling.ts [hunk 7]
```diff
@@ -17,7 +17,7 @@ import { getTokenInfo as defaultGetTokenInfo } from './oauth-tooling';
 import {
   AuthenticationMiddlewareOptions,
   ExtendedRequest,
-  onAuthorizationFailedHandler,
+  onAuthorizationFailedHandler as onAuthorizationFailedHandlerType,
```
Comment (ISO50) [inline src/express-tooling.ts:20]: I am not quite sure about the intent of this renaming, would the import type feature help here? 
 `import type { .... } from `

File: .eslintrc.js [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+    "no-multiple-empty-lines": "error",
+    "no-new-wrappers": "error",
+    "no-shadow": [
+      "error",
+      {
+        "hoist": "all"
+      }
+    ],
+    "no-trailing-spaces": "error",
+    "no-underscore-dangle":
```
Comment (Retro64) [inline .eslintrc.js:365]: switch off

File: src/oauth-tooling.ts (hunk 2 above)
Comment (Retro64) [reply src/oauth-tooling.ts:289]: was needed to avoid a possible null access in callee

File: test/unit/express-tooling/authentication-middleware.ts (hunk 4 above)
Comment (Retro64) [reply test/unit/express-tooling/authentication-middleware.ts:42]: how about dropping the curly braces?

**Comment by Retro64** [comment]
:+1:

**Comment by ISO50** [comment]
> > @bzums @ISO50 we should think about the version number before publishing
> 
> Imo we maybe do not even need to pulblish this as a new version since the compiled artifact will not differ at all. So merging would be sufficient. What do you think?

For me it depends whether the public API changed or not, did we changed types of public functions?

**Comment by Retro64** [comment]
:+1:

**Comment by gnuton** [comment]
If this is good to go can we get this merged? we have vulnerabilities we wanna get rid because of loadash.. if this is not ready could just we get some a minor release with up-to-date dependencies?

File: src/express-tooling.ts (hunk 7 above)
Comment (bzums) [reply src/express-tooling.ts:20]: I also do not see a reason to rename here

File: src/oauth-tooling.ts [hunk 9]
```diff
@@ -47,14 +47,14 @@ function createAuthCodeRequestUri(authorizationEndpoint: string,
                                   clientId: string,
                                   queryParams?: { [index: string]: string }): string {
 
-  const _queryParams = {
+  const extendedQueryParams = {
+    ...queryParams,
```
Comment (bzums) [inline src/oauth-tooling.ts:51]: This is potentially changing behaviour, I would like to undo this change.

**Comment by ISO50** [comment]
👍

**Comment by bzums** [comment]
👍

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,5 +1,9 @@
 # Changelog
 
+## `authmosphere 3.0.0`
+
+This release is probably non-breaking. However there are some changes to types due to an updated TypeScript version that we are not 100% sure about this. If you identify breaking behaviour/changes in this release, feel free to reach out us!
```
Comment (ISO50) [inline CHANGELOG.md:5]: > If you identify breaking behaviour/changes in this release, feel free to reach out us!

reach out to us

Comment (ISO50) [inline CHANGELOG.md:5]: > version that we are not 100%

version so we are not 100%

Comment (ISO50) [reply CHANGELOG.md:5]: > However there are some changes to types

Proposal:
However there are some types affected by changes

Comment (ISO50) [inline CHANGELOG.md:5]: > However there are some changes to types

Proposal:
However there are some types affected by changes

**Comment by bzums** [comment]
👍

**Comment by ISO50** [comment]
👍

**Comment by bzums** [comment]
👍

**Review by ISO50 — COMMENTED** [review]
LGTM

**Comment by ISO50** [comment]
👍
